### ISSUE TITLE
Custom http agent - reuse single tcp connection for multiple requests

### ISSUE THREAD

**Issue by mStirner** [nodejs/help#4149]
### Details

I try to reuse a single tcp connection for multiple requests.
The problem is, the code below throws a `ABORT_ERR`:

```
node:events:491
      throw er; // Unhandled 'error' event
      ^

AbortError: The operation was aborted
    at Duplexify.d._destroy (node:internal/streams/duplexify:359:13)
    at _destroy (node:internal/streams/destroy:102:25)
    at Duplexify.destroy (node:internal/streams/destroy:64:5)
    at Agent.<anonymous> (node:_http_agent:169:14)
    at Agent.emit (node:events:513:28)
    at Duplexify.onFree (node:_http_agent:383:11)
    at Duplexify.emit (node:events:513:28)
    at emitFreeNT (node:_http_client:794:16)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
Emitted 'error' event on Duplexify instance at:
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  code: 'ABORT_ERR'
}
```

The code is switching on/off some lights of my Raspberry ZigBee Gateway (RaspBee/Phoscon).
Both lights turn on, and then the error is thrown. The "Switch off" timeout is never fired.

For what i understand:
- The agent should schedule the requests to the available sockets.
- `maxSockets=1` tells the agent, to keep only one connection to `192.168.2.4`.
- So `agent.createConnection` is called for each requests (because then the socket should be destroyed after each request)

But that not the case. `createConnection` is called once, both lights turn on, the abort error is thrown.
Why `stream.destory` is not called?

---

Or in simpler words: How to reuse a single tcp socket for multiple http requests in the custom agent?

### Node.js version

16.19.1

### Example code

__index.js__
```js
const request = require("./request.js");

console.log("Started");


const agent = require("./http-agent.js");


console.log("Switch lights on");

request("http://192.168.2.4/api/BA3B2EDDA7/lights/4/state", {
    method: "PUT",
    agent,
    body: JSON.stringify({
        on: true
    })
});

request("http://192.168.2.4/api/BA3B2EDDA7/lights/1/state", {
    method: "PUT",
    agent,
    body: JSON.stringify({
        on: true
    })
});


setTimeout(() => {

    console.log("Switch lights off");

    request("http://192.168.2.4/api/BA3B2EDDA7/lights/4/state", {
        method: "PUT",
        agent,
        body: JSON.stringify({
            on: false
        })
    });

    request("http://192.168.2.4/api/BA3B2EDDA7/lights/1/state", {
        method: "PUT",
        agent,
        body: JSON.stringify({
            on: false
        })
    });

}, 3000);
```

__request.js__
```js
const url = require("url");

/**
 * Does a http request
 * @param {*} uri 
 * @param {*} options 
 * @param {*} cb 
 * 
 * @ignore
 * 
 * @returns {http.ClientRequest} https://nodejs.org/dist/latest-v16.x/docs/api/http.html#class-httpclientrequest
 */
function perform(uri, options, cb) {

    let { protocol } = new url.URL(uri);

    if (!["http:", "https:"].includes(protocol)) {
        throw new Error(`Unspported protocol "${protocol.slice(0, -1)}`);
    }

    // NOTE: Automaticly set keep alive header when agent is passed?
    if (options?.agent && !options?.setKeepAliveHeader) {
        options.setKeepAliveHeader = true;
    }

    if (options?.setKeepAliveHeader) {
        options.headers = {
            ...options?.headers,
            "Connection": "Keep-Alive"
        };
    }

    let request = require(protocol.slice(0, -1)).request(uri, options, (res) => {

        let chunks = [];

        res.on("data", (chunk) => {
            chunks.push(chunk);
        });

        res.on("error", cb);

        res.on("end", () => {

            let body = Buffer.concat(chunks);

            if (res.headers["content-type"] && res.headers["content-type"].includes("application/json")) {
                body = JSON.parse(body);
            }

            cb(null, {
                headers: res.headers,
                status: res.statusCode,
                body
            });

        });

    });


    request.on("error", (err) => {
        cb(err);
    });


    if (options.callEnd) {
        request.end(options.body);
    }
    //request.write(options.body + "\r\n");

    return request;

}


/**
 *  @function request
 * Does a http/https request
 *
 * @param {String} uri 
 * @param {Object} options 
 * @param {Function} cb Callback

 * @returns {http.ClientRequest} https://nodejs.org/dist/latest-v16.x/docs/api/http.html#class-httpclientrequest
 */
module.exports = function request(uri, options, cb) {

    if (!cb && options instanceof Function) {
        cb = options;
        options = {};
    }

    if (!cb) {
        cb = () => { };
    }

    options = Object.assign({
        method: "GET",
        body: "",
        followRedirects: true,
        callEnd: true,
        setKeepAliveHeader: false
    }, options);


    return perform(uri, options, (err, result) => {
        if (err) {

            cb(err);

        } else {

            if (options.followRedirects && result.status >= 300 && result.status < 400) {

                perform(result.headers.location, options, cb);

            } else {

                cb(null, result);

            }

        }
    });

};
```
__http-agent.js__
```js
const { Agent } = require("http");
const { createConnection, connect } = require("net");
const { Duplex, PassThrough } = require("stream");


// maxSockets = 1, creatConnection is only called once!
const agent = new Agent({
    maxSockets: 1
});


const socket = createConnection({
    host: "192.168.2.4",
    port: 80
});



agent.createConnection = (options) => {

    console.log("Create connection", options.href);

    let readable = new PassThrough();
    let writable = new PassThrough();


    let stream = new Duplex.from({
        readable,
        writable
    });

    socket.pipe(readable);
    writable.pipe(socket);

    return stream;

};

module.exports = agent;
```

### Operating system

Ubuntu 18.04.6 LTS

### Scope

Code/http agent

### Module and version

Not applicable.

**Comment by preveen-stack** [comment]
```
from flask import Flask

app = Flask(__name__)

@app.route("/")
def hello_world():
    return "<p>Hello, World!</p>"
```
started a flask server on the local system for simulation
```
$ flask --app flask_qs.py run --debug
 * Serving Flask app 'flask_qs.py'
 * Debug mode: on
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
 * Restarting with stat
 * Debugger is active!
 * Debugger PIN: 977-055-346

```

adjusted the host name in the index.js and http-agent.js to connect on 127.0.0.1 port 5000
```
Started
Switch lights on
Create connection http://127.0.0.1:5000/api/BA3B2EDDA7/lights/4/state
Create connection http://127.0.0.1:5000/api/BA3B2EDDA7/lights/4/state
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: This socket has been ended by the other party
    at Socket.writeAfterFIN [as write] (node:net:497:14)
    at PassThrough.ondata (node:internal/streams/readable:766:22)
    at PassThrough.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
    at Readable.push (node:internal/streams/readable:234:10)
    at node:internal/streams/transform:182:12
    at PassThrough._transform (node:internal/streams/passthrough:46:3)
    at Transform._write (node:internal/streams/transform:175:8)
Emitted 'error' event on Socket instance at:
    at Socket.onerror (node:internal/streams/readable:785:14)
    at Socket.emit (node:events:525:35)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  code: 'EPIPE'
}

Node.js v18.7.0

```

changed maxsockets to 2 in http-agent.js 

```

Started
Switch lights on
Create connection http://127.0.0.1:5000/api/BA3B2EDDA7/lights/4/state
Create connection http://127.0.0.1:5000/api/BA3B2EDDA7/lights/1/state
Switch lights off
Create connection http://127.0.0.1:5000/api/BA3B2EDDA7/lights/4/state
Create connection http://127.0.0.1:5000/api/BA3B2EDDA7/lights/1/state
```
no error was observed

**Comment by preveen-stack** [comment]
Can you share the link to the docs

**Comment by preveen-stack** [comment]
I will look through the docs.
meanwhile check this issue https://github.com/node-fetch/node-fetch/issues/1153

**Comment by preveen-stack** [comment]
Can you change  the keep alive option to true and do a test to see if any effect is there

**Comment by preveen-stack** [comment]
I tried with the keepAlive : true option, but getting the following error
```
Started
Switch lights on
Create connection http://127.0.0.1:8000/api/BA3B2EDDA7/lights/4/state
keepAlive: true
node:_http_agent:492
  socket.setKeepAlive(true, this.keepAliveMsecs);
         ^

TypeError: socket.setKeepAlive is not a function
    at Agent.keepSocketAlive (node:_http_agent:492:10)
    at Agent.<anonymous> (node:_http_agent:185:15)
    at Agent.emit (node:events:513:28)
    at Duplexify.onFree (node:_http_agent:386:11)
    at Duplexify.emit (node:events:513:28)
    at emitFreeNT (node:_http_client:769:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21)

Node.js v18.7.0

```

**Comment by mStirner** [comment]
Seems like with monkey patching these methods onto the duplex stream created in `.createConnection()`:

```js
const { Agent } = require("http");
const { createConnection, connect } = require("net");
const { Duplex, PassThrough } = require("stream");


// maxSockets = 1, creatConnection is only called once!
const agent = new Agent({
    keepAlive: true,
    maxSockets: 1
});


const socket = createConnection({
    host: "192.168.2.4",
    port: 80
});



agent.createConnection = (options) => {

    console.log("Create connection", options.href);

    let readable = new PassThrough();
    let writable = new PassThrough();


    let stream = new Duplex.from({
        readable,
        writable
    });

    stream.ref = (...args) => { console.log("ref", args); };
    stream.unref = (...args) => { console.log("unref", args); };    
    stream.setKeepAlive = (...args) => { console.log("setKeepAlive", args); };
    stream.setTimeout = (...args) => { console.log("setTimeout", args); };

    socket.pipe(readable);
    writable.pipe(socket);

    return stream;

};

module.exports = agent;
```
Output:
```
Started
Switch lights on
Create connection http://192.168.2.4/api/BA3B2EDDA7/lights/4/state
Connectected to tcp://192.168.2.4:80
setKeepAlive [ true, 1000 ]
unref []
setTimeout [ 0 ]
Switch lights off
ref []
setKeepAlive [ true, 1000 ]
unref []
setTimeout [ 0 ]
```

But im not sure why to set `keepAlive=true`, and monkey patch the `setKeepAlive`, `setTimeout`, `ref` and `unref`, methods is necessary when `maxSockets=1` indicates that the created socket/stream should be re-used.

A data-flow diagram how the adapter works under the hood would be nice to understand that better.
Some like for the stream back pressure found here: https://nodejs.org/en/docs/guides/backpressuring-in-streams#lifecycle-of-pipe

**Comment by preveen-stack** [comment]
How can we avoid putting host and port in http-agent.js

**Comment by mStirner** [comment]
You can move the tcp socket creation inside the `agent.createConnection` function (where it should be). I have put it outside to demonstrate it. Then you can just extract host/port from the options object and pass that to `net.createConnection()`.

NOTE About the code i wrote:
I have a edge case where i cant use a tcp socket created from `net.createConnection()` to communicate with a HTTP API/Device.
Thats why i created a separate duplex stream, and patch the needed functions. 

If you need the same functionality above, but dont need to workaround the tcp socket, just create a custom agent with `keepAlive=true` & `maxSockets=1`. No need to override the `agent.createConnection` function.

**Comment by preveen-stack** [comment]
BTW, what are you using on the server side?

**Comment by preveen-stack** [comment]
Curious about  your rest api server at 192.168.2.4 running on which platform

**Comment by mStirner** [comment]
Thats a ZigBee gateway based on a raspberry pi and "add-on" module from "Dresden Elektronik": https://phoscon.de/en/raspbee

**Comment by mStirner** [comment]
I found a issue with the approach above:

When a server sets the connection header to "close", the agent breaks and does nothing more.
For example, i have a shelly device, with i want to fetch three routes: `/status`, `/shell` and  `/settings`.

The first request is done, the following not.

Minimal reproduce example with the request code from above:
```js
const { Agent } = require("http");
const { createConnection } = require("net");
const { Duplex, PassThrough } = require("stream");

const request = require("./request.js");

(async () => {

    // maxSockets = 1, creatConnection is only called once!
    const agent = new Agent({
        keepAlive: true,
        maxSockets: 1
    });


    agent.createConnection = (options) => {

        console.log("Create connection", options.href);

        const socket = createConnection({
            host: "192.168.2.115",
            port: 80
        });

        socket.setKeepAlive(true, 5000);

        socket.on("connect", () => {
            console.log("Connectected to tcp://192.168.2.115:80");
            socket.emit("connected");
        });

        socket.on("close", () => {
            console.log("TCP socket closed");
        });

        let readable = new PassThrough();
        let writable = new PassThrough();


        let stream = new Duplex.from({
            readable,
            writable
        });

        stream.destroy = () => {
            console.log("Destory called", stream);
        }

        // do this when interface attached?
        // and vice verse when disconnected?
        //stream.connecting = false;
        //stream.emit("connected");

        stream.unref = (...args) => {
            console.log("unref", args);
        };
        stream.ref = (...args) => {
            console.log("ref", args);
        };
        stream.setKeepAlive = (...args) => {
            console.log("setKeepAlive", args);
        };
        stream.setTimeout = (...args) => {
            console.log("setTimeout", args);
        };



        socket.pipe(readable, { end: false });
        writable.pipe(socket, { end: false });

        return stream;
        //return socket;

    };

    //agent.createConnection = createConnection;




    let settings = await new Promise((resolve, reject) => {
        request("http://192.168.2.115/settings", {
            agent
        }, (err, result) => {
            if (err) {
                reject(err);
            } else {
                console.log("resut", result)
                resolve(result.body)
            }
        });
    });

    console.log("Settings fetched")

    new Promise((resolve) => { setTimeout(resolve, 1000); });


    let shelly = await new Promise((resolve, reject) => {
        request("http://192.168.2.115/shelly", {
            agent
        }, (err, result) => {
            if (err) {
                reject(err);
            } else {
                resolve(result.body)
            }
        });
    });

    console.log("Shelly fetched")

    new Promise((resolve) => { setTimeout(resolve, 1000); });


    let status = await new Promise((resolve, reject) => {
        request("http://192.168.2.115/status", {
            agent
        }, (err, result) => {
            if (err) {
                reject(err);
            } else {
                resolve(result.body)
            }
        });
    });

    console.log("shelly fetched")

    console.log(settings, shelly, status)

})();
```

Output:
```
Perform request to http://192.168.2.115/settings
Set keep alive header http://192.168.2.115/settings
Create connection http://192.168.2.115/settings
Connectected to tcp://192.168.2.115:80
Settings fetched
Perform request to http://192.168.2.115/shelly
Set keep alive header http://192.168.2.115/shelly
TCP socket closed
```

For me it seems like overriding the the `.destroy` function is causing this.
When a "original" net socket is returned, the requests work. As soon i override the `.destory` function for said socket, the issue happens.

---

Dear node.js team, is there somewhere a flow diagram about the agent?

**Comment by preveen-stack** [comment]
May be the following from the [http](https://nodejs.org/api/http.html#class-httpagent) doc can shed some light on understanding agent behaviour

> n Agent is responsible for managing connection persistence and reuse for HTTP clients. It maintains a queue of pending requests for a given host and port, reusing a single socket connection for each until the queue is empty, at which time the socket is either destroyed or put into a pool where it is kept to be used again for requests to the same host and port. Whether it is destroyed or pooled depends on the keepAlive [option](https://nodejs.org/api/http.html#new-agentoptions).
> 
> Pooled connections have TCP Keep-Alive enabled for them, but servers may still close idle connections, in which case they will be removed from the pool and a new connection will be made when a new HTTP request is made for that host and port. Servers may also refuse to allow multiple requests over the same connection, in which case the connection will have to be remade for every request and cannot be pooled. The Agent will still make the requests to that server, but each one will occur over a new connection.
> 
> When a connection is closed by the client or the server, it is removed from the pool. Any unused sockets in the pool will be unrefed so as not to keep the Node.js process running when there are no outstanding requests. (see [socket.unref()](https://nodejs.org/api/net.html#socketunref)).
> 
> It is good practice, to [destroy()](https://nodejs.org/api/http.html#agentdestroy) an Agent instance when it is no longer in use, because unused sockets consume OS resources.

**Comment by mStirner** [comment]
@preveen-stack  I understand that. But what i have noticed is that the connection is not "removed"/freed from the pool after the request is done. It just stay in a "zombie" state. And because of `maxSockets=1`, no new connections are created.

Simple server/example:

_server.js_
```js
const { createServer } = require("http");
const os = require("os");

const server = createServer((req, res) => {

    console.log(`[${req.method}] ${req.url}`);

    //res.socket.setKeepAlive(true, 5000);

    res.setHeader("content-type", "application/json");
    res.setHeader("connection", "close");

    res.write(JSON.stringify({
        hostname: os.hostname(),
        timestamp: Date.now(),
        url: req.url,
        method: req.method
    }));

    res.end();

});

server.listen(8080, (err) => {
    console.log(err || "HTTP Server listening on http://127.0.0.1:8080");
});
```

_index.js_
```js
const request = require("./request.js");

console.log("Started");
const agent = require("./http-agent.js");
const BASE_URL = "http://127.0.0.1:80"


console.log("Switch lights on");

request(`http://${BASE_URL}/shelly`, {
    agent
}, (err, { body }) => {
    console.log("/shelly", err || body);
});

request(`http://${BASE_URL}/settings`, {
    agent
}, (err, { body }) => {
    console.log("/settings", err || body);
});

request(`http://${BASE_URL}/status`, {
    agent
}, (err, { body }) => {
    console.log("/status", err || body);
});
```

_agent.js_
```js
const { Agent } = require("http");
const { createConnection } = require("net");
const { Duplex, PassThrough } = require("stream");

const socket = createConnection({
    host: "127.0.0.1",
    port: 8080
});

socket.on("connect", () => {
    console.log("Connectected to tcp://127.0.0.1:8080");
});

socket.on("close", () => {
    console.log("TCP socket closed");
});


// maxSockets = 1, creatConnection is only called once!
const agent = new Agent({
    keepAlive: true,
    maxSockets: 1
});

agent.createConnection = (options) => {

    console.log("Create connection", options.href);

    let readable = new PassThrough();
    let writable = new PassThrough();


    let stream = new Duplex.from({
        readable,
        writable
    });

    stream.destroy = (...args) => {

        console.log("Destory called", args);

        socket.unpipe();

    };

    stream.ref = (...args) => { console.log("ref", args); };
    stream.unref = (...args) => { console.log("unref", args); };
    stream.setKeepAlive = (...args) => { console.log("setKeepAlive", args); };
    stream.setTimeout = (...args) => { console.log("setTimeout", args); };

    socket.pipe(readable);
    writable.pipe(socket);

    return stream;
    //return socket;

};

module.exports = agent;
```

Output:
```
Started
Switch lights on
Create connection http://http//127.0.0.1:80/shelly
Connectected to tcp://127.0.0.1:8080
/shelly {
  hostname: 'stirner',
  timestamp: 1682406465230,
  url: '//127.0.0.1:80/shelly',
  method: 'GET'
}
Destory called []
Destory called []
TCP socket closed

```

The first request is done, for the other two (i assume) `.destroy` is called on the agent duplex stream/socket.
(Note: the tcp socket is also closed, but this is currently no problem, because `agent.createConnection` is not called)

I dont understand why, the first request is done and for the other 2 no new connection is created. (or freed at least)
What do i need to do, to tell the agent "yup, socket is destroyed, call `agent.createConnection` again for the pending requests".

**Comment by mStirner** [comment]
Setting `end=false` on the piping, does not call the `.destroy()`  method`:

```js
socket.pipe(readable, { end: false });
writable.pipe(socket, { end: false });
```

```
Started
Switch lights on
Create connection http://http//127.0.0.1:80/shelly
Connectected to tcp://127.0.0.1:8080
/shelly {
  hostname: 'stirner',
  timestamp: 1682407534690,
  url: '//127.0.0.1:80/shelly',
  method: 'GET'
}
TCP socket closed
```

No sure how to signal the agent he should keep processing the pending requests.

**Comment by preveen-stack** [comment]
```
const http = require('http');

// Create an instance of the Agent class
const agent = new http.Agent({
  keepAlive: true,
  maxSockets: 1
});

// Create options object with hostname and agent properties
const options = {
  hostname: 'example.com',
  agent: agent
};

// Make multiple requests using the same agent instance
http.get(options, (res) => {
  console.log('Response received');
}).on('error', (err) => {
  console.error(err);
});

http.get(options, (res) => {
  console.log('Response received');
}).on('error', (err) => {
  console.error(err);
});

http.get(options, (res) => {
  console.log('Response received');
}).on('error', (err) => {
  console.error(err);
});

```

**Comment by mStirner** [comment]
@preveen-stack  You didnt use a custom `agent.createConnection` function. With the vanilla function or create socket from `net.createConnection` no problem. But when you return a duplex stream, and  not a `net.Socket` it breaks.

**Comment by preveen-stack** [comment]
I guess we have to do synchronous call, however the http apis are inherently async
```
const http = require('http');

// Create an instance of the Agent class
const agent = new http.Agent({
  keepAlive: true,
  maxSockets: 1 
});

const options = {
  hostname: '127.0.0.1',
  port: 8000,
  path: '/api',
  method: 'GET',
  agent : agent
};

const hostkey = `${options.hostname}.${options.port}`;
console.log(hostkey)

function makeRequest(requestNumber) {
  const req = http.request(options);

  const data = [];

  req.on('response', (res) => {
    res.on('data', (chunk) => {
      data.push(chunk);
    });

    res.on('end', () => {
      const response = Buffer.concat(data).toString();
      console.log("socket count: " + agent.totalSocketCount);
      if (requestNumber < 5) {
        makeRequest(requestNumber + 1);
      }
    });
  });

  req.end();
}

makeRequest(1);
```
```
Response from request 1: Success
socket count: 1
Response from request 2: Success
socket count: 1
Response from request 3: Success
socket count: 1
Response from request 4: Success
socket count: 1
Response from request 5: Success
socket count: 1
```

Serverlog
```
2023-04-26T05:15:12.954Z Listening on port 8000...
2023-04-26T05:17:41.279Z New connection 1682486261279-378361
2023-04-26T05:17:41.280Z Incoming request
2023-04-26T05:17:42.298Z Response is sent
2023-04-26T05:17:42.301Z Incoming request
2023-04-26T05:17:43.302Z Response is sent
2023-04-26T05:17:43.303Z Incoming request
2023-04-26T05:17:44.304Z Response is sent
2023-04-26T05:17:44.305Z Incoming request
2023-04-26T05:17:45.305Z Response is sent
2023-04-26T05:17:45.306Z Incoming request
2023-04-26T05:17:46.308Z Response is sent
2023-04-26T05:17:46.312Z Connection closed 1682486261279-378361
```
As you can see only one connection is made; If you turn off keep alive then you can see a connection being made and closed for each request. 

May be we can use async.series to get make synchronous http calls, but not sure

**Comment by mStirner** [comment]
I close this for now, because that has nothing more to do with the original issue, which works now with `maxSockets=1` & `setKeepAlive=true`
